core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/GreedyExecutionPlanGenerator.scala

```diff
@@ -0,0 +1,394 @@
+      // return all links for materialization replacement
+      return Some(
+        upstreamPhysicalOpIds
+          .flatMap { upstreamPhysicalOpId =>
+            physicalPlan.getLinksBetween(upstreamPhysicalOpId, physicalOpId)
+          }
+      )
+    }
+    None
+  }
+
+  private def handleDependentLinks(
+      physicalOpId: PhysicalOpIdentity,
+      regionDAG: DirectedAcyclicGraph[Region, RegionLink]
+  ): Option[List[PhysicalLink]] = {
+    // for operators like HashJoin that have an order among their blocking and pipelined inputs
+    physicalPlan
+      .getOperator(physicalOpId)
+      .getInputLinksInProcessingOrder
+      .sliding(2, 1)
+      .foreach {
+        case List(prevLink, nextLink) =>
+          // Create edges between regions
+          val regionLinks = createLinks(prevLink.fromOp.id, nextLink.fromOp.id, regionDAG)
+          // Attempt to add edges to regionDAG
+          try {
+            regionLinks.foreach(link => regionDAG.addEdge(link.fromRegion, link.toRegion, link))
+          } catch {
+            case _: IllegalArgumentException =>
+              // adding the edge causes cycle. return the link materialization replacement
+              return Some(List(nextLink))
+          }
+      }
+    None
+  }
+
+  /**
+    * This function creates and connects a region DAG while conducting materialization replacement.
+    * It keeps attempting to create a region DAG from the given PhysicalPlan. When failed, a list
+    * of PhysicalLinks that causes the failure will be given to conduct materialization replacement,
+    * which changes the PhysicalPlan. It keeps attempting with the updated PhysicalPLan until a
+    * region DAG is built after connecting materialized pairs.
+    *
+    * @return a fully connected region DAG.
+    */
+  private def createRegionDAG(): DirectedAcyclicGraph[Region, RegionLink] = {
+    val matReaderWriterPairs =
+      new mutable.HashMap[PhysicalOpIdentity, PhysicalOpIdentity]()
+    @tailrec
+    def recConnectRegionDAG(): DirectedAcyclicGraph[Region, RegionLink] = {
+      tryConnectRegionDAG() match {
+        case Left(dag) => dag
+        case Right(links) =>
+          links.foreach { link =>
+            physicalPlan = replaceLinkWithMaterialization(link, matReaderWriterPairs)
+          }
+          recConnectRegionDAG()
+      }
+    }
+
+    // the region is partially connected successfully.
+    val regionDAG: DirectedAcyclicGraph[Region, RegionLink] = recConnectRegionDAG()
+
+    // the update physical plan's all source operators should have 0 input ports
+    physicalPlan.getSourceOperatorIds.foreach { sourcePhysicalOpId =>
+      assert(physicalPlan.getOperator(sourcePhysicalOpId).inputPorts.isEmpty)
+    }
+    // the update physical plan's all sink operators should have 0 output ports
+    physicalPlan.getSinkOperatorIds.foreach { sinkPhysicalOpId =>
+      assert(physicalPlan.getOperator(sinkPhysicalOpId).outputPorts.isEmpty)
+    }
```

Comment:
I suggest letting the compiler do this check after the execution plan generation.

Reply:
moved back to compiler